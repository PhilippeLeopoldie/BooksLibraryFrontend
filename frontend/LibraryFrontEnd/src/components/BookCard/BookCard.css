.bookcard--grid {
    border: 1px rgba(0, 0, 0, 0.2) solid;
    display: grid;
    grid-template-rows: 68% 18% 14%;
    width: 18rem;
    min-width: 18rem;
    height: 29rem;
    min-height: 29rem;
    margin: 0 1rem 2rem 1rem;
    background-color: var(--backgroundcolor-white);
    border-radius: 10px;
    box-shadow: var(--cardboxshadow-lightmode);
}

.bookcard--grid:hover {
    box-shadow: var(--cardboxshadow-hover-lightmode);
}

.bookcard--black:hover {
    border: 1px var(--cardborder-hover) solid;
    box-shadow: var(--cardboxshadow-hover-darkmode);
}

.bookcard--black {
    background-color: inherit;
    box-shadow: var(--cardboxshadow-darkmode);
    & .booktitle, & .bookauthor {
    color: var(--fontcolor-dark);
    }
}


