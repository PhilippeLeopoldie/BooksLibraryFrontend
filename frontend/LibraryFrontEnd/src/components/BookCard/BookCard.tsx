import "./BookCard.css";
import { BookZoomCover } from "./BookCover/BookZoomCover/BookZoomCover";
import { BookCover } from "./BookCover/BookCover";
import { BookBody } from "./BookBody/BookBody";
import { BookType, ReviewType } from "../../constants/types";
import { OpinionCreate } from "../Opinions/OpinionCreate/OpinionCreate";
import { OpinionList } from "../Opinions/OpinionList/OpinionList";
import { ThemeContext } from "../../App/App";
import { BOOK_BY_BOOKID_URL } from "../../constants/api";
import { useContext, useEffect, useState } from "react";
import { BookFooter } from "./BookFooter/BookFooter";

type CardState = {
  updatedBook: BookType,
  curentView: string, 
  opinionCreated: boolean,
  averageRateClick: boolean,
  reviewList: ReviewType[],
}

export const BookCard = ({ book }: { book: BookType }) => {
  const theme = useContext(ThemeContext);
  const [card, setCard] = useState <CardState>({
    updatedBook : book  ,
    curentView : "bookPresentation",
    opinionCreated: false,
    averageRateClick: false,
    reviewList:[],
  });

  const handleAverageRateClick = () => {
    setCard(prevBookCard => ({...prevBookCard, averageRateClick: !card.averageRateClick}));
  }


  // used in handleOpinionList to render <OpinionList/> when 'curentView'=== 'opinionList'
  const toggleOpinionList = () => {
    card.curentView === "bookPresentation"
      ? setCard(prevView => ({...prevView,curentView:"opinionList"}))
      : setCard(prevView => ({...prevView, curentView:"bookPresentation"}));
    setCard(prevAverageClick => ({...prevAverageClick, averageRateClick: false}));
  };

  const handleOpinionList = (reviews: ReviewType[]) => {
    toggleOpinionList();
    setCard(prevBookCard => ({...prevBookCard, reviewList: reviews}));
  };

  const handleCreatedOpinion = (value: boolean) => {
    setCard(prevOpinionCreatedValue => ({...prevOpinionCreatedValue, opinionCreated:value}));
  };
  const toggleCreateOpinion = (view: string) => {
    setCard(prevView => ({...prevView, curentView: view}))
  };

  const fetchBook = async () => {
    try {
      const bookResponse: Response = await fetch(
        `${BOOK_BY_BOOKID_URL}${book?.id}`
      );
        if (bookResponse.status === 200) {
            const bookResponseData = await bookResponse.json();
          setCard(prevState => ({ ...prevState, updatedBook: bookResponseData.book, opinionCreated: false }));
      } else if (bookResponse.status === 404) {
        console.log(bookResponse);
      }
    } catch (error) {
      console.error(`Error fetching bookId:${book?.id}`, error);
    }
  };
  // when a new opinion is posted ,fetch the book with the new averageRate updated
  useEffect(() => {
      card.opinionCreated === true && fetchBook();
  }, [card.opinionCreated]);

  //when 'displayContent' is rendered it will render either <OpinionList/> , <bookPresentation/> or <createOpinion/>
  let displayedContent: JSX.Element = <></>;
  if (book && card.curentView === "bookPresentation") {
    const bookFooter = {
      toggleCreateOpinion: toggleCreateOpinion,
      handleOpinionList : handleOpinionList,
      handleAverageClick: handleAverageRateClick,
      averageClick: card.averageRateClick,
    }
    displayedContent = (
      <section className={"bookcard--grid bookcard--" + theme}>
        <BookCover book={book} />
        <BookBody  book = {book} />
        <BookFooter bookFooter={bookFooter} updatedBook={card.updatedBook}/>
        
      </section>
    );
  }
  if (card.curentView === "opinionList") {
    displayedContent = (
      <OpinionList
        opinions={card.reviewList}
        displayReviews={() => toggleOpinionList()}
        book={book}
      />
    );
  }
  if (card.curentView === "createOpinion") {
    displayedContent = (
      <OpinionCreate
        book={book && book}
        toCreate={toggleCreateOpinion}
        created={handleCreatedOpinion}
      />
    );
  }

  return <>{displayedContent}</>;
};
