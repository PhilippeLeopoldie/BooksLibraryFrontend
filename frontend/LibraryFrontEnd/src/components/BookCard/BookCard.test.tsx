import { render, screen } from "@testing-library/react";
import { BookCard } from "./BookCard";
import "@testing-library/jest-dom";

const actualBookData = {
  id: 1,
  title: "The Great Gatsby",
    author: "F. Scott Fitzgerald",
    description: "The story of the mysteriously wealthy Jay Gatsby and his love for the beautiful Daisy Buchanan.",
  averageRate: 3,
};

describe("", () => {
  test("snapchot: renders book title and author", () => {
    // Arrange and Act
    const { asFragment } = render(
      <BookCard book={actualBookData}  />
    );

    //expect (snapshot)
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders book title and author", () => {
    // Arrange
    render(<BookCard book={actualBookData}  />);

    // Act
    const titleElement = screen.getByText("The Great Gatsby");
    const authorElement = screen.getByText("by: F. Scott Fitzgerald");

    // Expect
    expect(titleElement).toBeInTheDocument();
    expect(authorElement).toBeInTheDocument();
    expect(titleElement).toHaveClass("booktitle");
    expect(authorElement).toHaveClass("bookauthor");
  });

/*   test("renders delete img with the correct alt text and ClassName", () => {
    // Arrange
    const onDelete = jest.fn();
    const { getByAltText } = render(
      <Book book={actualBookData} onDelete={onDelete} />
    );

    // Act
    const trashImage = getByAltText("delete");

    //Expect
    expect(trashImage).toBeInTheDocument();
    expect(trashImage).toHaveClass("bookcard--iconeTrash");
  }); */
});
