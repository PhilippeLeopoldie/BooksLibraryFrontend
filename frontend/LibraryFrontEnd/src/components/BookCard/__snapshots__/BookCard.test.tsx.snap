// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` snapchot: renders book title and author 1`] = `
<DocumentFragment>
  <div
    class="bookcard--grid bookcard--natural"
  >
    <img
      alt="bookImage"
      class="boocard__Image-zoomOut boocard__Image"
    />
    <div
      class="bookcard__body"
    >
      <h3
        class="booktitle"
        title="Book Title"
      >
        The Great Gatsby
      </h3>
      <h3
        class="bookauthor"
      >
        by: F. Scott Fitzgerald
      </h3>
    </div>
    <footer
      class="bookcard__footer--flex"
    >
      <div
        class="bookcard__footer-average-rate--flex"
      >
        <div
          class="bookcard__footer-average-rate--flex"
        >
          <a
            class="bookcard__footer-average-rate"
          >
            3/5
          </a>
          <div
            class="rate_star"
          >
             â˜…
          </div>
        </div>
      </div>
      <button
        class="button bookFooter__RateButton"
      >
        Rate this book
      </button>
    </footer>
  </div>
</DocumentFragment>
`;
